tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,337 @@
+    """Gaussian noise distribution"""
+    if self._stateless:
+      offset_seed = self._seed_offset + kwargs.get('seed', 0)
+      kwargs['seed'] = tf.cast(
+        tf.stack([offset_seed, step], tf.int32)
+      )
+      return tf.random.stateless_normal(*args, **kwargs)
+    return tf.random.normal(*args, **kwargs)
+
+  def _get_grow_grads(self, optimizer, step, mask, grad):
+    """Grow connections baased on gradient information.
+    """
+    grad_scores = tf.math.abs(grad)
+    return grad_scores
+
+  def _get_drop_weights(self, optimizer, step, mask, weight, noise_std=0):
+    """
+    Drop connections based on weight magnitude.
+    """
+    masked_weight = mask * weight
+    weight_scores = tf.math.abs(masked_weight)
+    # add noise to break ties, although the possibility is very low
+    if noise_std != 0:
+      weight_scores.assign_add(self._random_normal(
+        weight_scores.shape, stddev=noise_std, dtype=weight_scores.dtype,
+        seed=(hash(weight.name + 'drop'))))
+    return weight_scores
+
+  def _reset_momentum(self, optimizer, weight, new_connections):
+    """Zeros out optimizer slots whose connections have been recovered."""
+    for slot_name in optimizer._optimizer.get_slot_names():
+      # reset aggregated momentum variables to 0
+      opt_var = optimizer._optimizer.get_slot(weight, slot_name)
+      new_values = tf.where(new_connections,
+                            tf.zeros_like(opt_var), opt_var)
+      opt_var.assign(new_values)
+
+  def _generic_top_k(self, scores, mask, n_new, n_total):
+     # sort the entire array so that it can be constant for TPUs
+    _, sorted_idx = tf.math.top_k(
+      tf.reshape(scores, (-1), k=n_total)
+    )
+    expanded_sorted_idx = tf.expand_dims(sorted_idx, 1)
+    new_values = tf.where(
+      tf.range(n_total) < n_new,
+      tf.ones_like(sorted_idx, dtype=mask.dtype),
+      tf.zeros_like(sorted_idx, dtype=mask.dtype)
+    )
+    updated_mask = tf.scatter_nd(expanded_sorted_idx, new_values, new_values.shape)
+
+    return updated_mask
+
+  def _get_new_connections(self, reinit_when_same, grown_mask_reshaped, mask):
+    if reinit_when_same:
+      new_connections = tf.math.equal(grown_mask_reshaped, 1)
+    else:
+      new_connections = tf.math.logical_and(
+        tf.math.equal(grown_mask_reshaped, 1), tf.math.equal(mask, 0)
+      )
+    return new_connections
+
+  def _update_mask(self, step, update_fraction, mask, weight, grad):
+    """Called by _maybe_update_block_mask.
+
+    Updates mask based on weight and grad information.
+    """
+    # compute the top k magnitudes then update the current mask
+    drop_scores = self._get_drop_weights(optimizer, step, mask, weight, noise_std=self._noise_std)
+    # need access to exactly which entries are growing to zero out optimizer slot
+    grow_scores = self._get_grow_grads(optimizer, step, mask, grad)
+    n_total = tf.size(drop_scores)
+    n_ones = tf.cast(tf.reduce_sum(mask), dtype=tf.int32)
```

Comment:
It's probably way more efficient to compute this from a sparsity float at the start rather than trying to reduce the mask every iteration.

This could be an argument for taking a sparsity float at the start instead of an initializer. (Though there are valid arguments for taking an initializer and not a sparsity float too.)